juju/charm.py

```diff
@@ -82,23 +82,33 @@ class CharmBase(Object):
 
     on = CharmEvents()
 
-    def __init__(self, framework, key, metadata, charm_dir):
+    def __init__(self, framework, key):
         super().__init__(framework, key)
-        self.metadata = metadata
-        self.charm_dir = charm_dir
 
-        for relation_name in self.metadata.relations:
+        for relation_name in self.meta.relations:
             relation_name = relation_name.replace('-', '_')
             self.on.define_event(f'{relation_name}_relation_joined', RelationJoinedEvent)
             self.on.define_event(f'{relation_name}_relation_changed', RelationChangedEvent)
             self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
             self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
 
-        for storage_name in metadata.storage:
+        for storage_name in self.meta.storage:
             storage_name = storage_name.replace('-', '_')
             self.on.define_event(f'{storage_name}_storage_attached', StorageAttachedEvent)
             self.on.define_event(f'{storage_name}_storage_detaching', StorageDetachingEvent)
 
+    @property
+    def charm_dir(self):
+        return self.framework.charm_dir
+
+    @property
+    def meta(self):
+        return self.framework.meta
+
+    @property
+    def model(self):
+        return self.framework.model
```

Comment:
I added the properties just as convenience accessors since I assumed devs would be more typically working with the charm object rather than the framework, but as I think about parts, they would likely only interact with the framework, and I definitely see your point about not wanting to have two "official" locations to access these.  +1 I'll remove them.